src/traversals/dfs.jl

```diff
@@ -5,12 +5,23 @@
     is_cyclic(g)
 
 Return `true` if graph `g` contains a cycle.
-
+  
 ### Implementation Notes
-Uses DFS.
+Undirect graph: for each connected component of the graph, the algorithm counts how many edges are in surplussurplus to a tree with the same number of nodes.
+Directed graph: uses DFS.
 """
 function is_cyclic end
-@traitfn is_cyclic(g::::(!IsDirected)) = ne(g) > 0
+@traitfn function is_cyclic(g::::(!IsDirected))
+    b=false
+    for vcomponent in connected_components(g)
+        sg, _ = induced_subgraph(g,vcomponent)
```

Comment:
Is there a way to do this without calling `induced_subgraph`? Judging by the [implementation](https://github.com/JuliaGraphs/Graphs.jl/blob/826b8feac41a2164b05a62296a1f0ac5158d703d/src/operators.jl#L639) this function might lead to significant overhead, so I'm wondering if a more efficient option exists

Reply:
I found a way. This was the performance before:
```Julia
julia> @benchmark(is_cyclic(gcyclic))
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.523 μs … 437.362 μs  ┊ GC (min … max):  0.00% … 99.39%
 Time  (median):     1.623 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   2.018 μs ±  10.616 μs  ┊ GC (mean ± σ):  13.89% ±  2.63%

    ▇█▅▁                                                       
  ▂▇████▇▆▅▄▃▃▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  1.52 μs         Histogram: frequency by time        2.58 μs <

 Memory estimate: 3.23 KiB, allocs estimate: 36.
 ```

With the following new version (count edges without create the subgraph):
```Julia
@traitfn function is_cyclic(g::::(!IsDirected))
    b=false
    for vcomponent in connected_components(g)
        ne=0
        l=length(vcomponent)  
        for i in 1:l
            for j in i+1:l
                if has_edge(g,vcomponent[i],vcomponent[j])
                    ne+=1
                end
            end
        end
        if (ne-(l-1)) > 0
            b = true
            break
        end
    end
    return b
end
```

The performance results:
```Julia
@benchmark(is_cyclic(gcyclic))
BenchmarkTools.Trial: 10000 samples with 160 evaluations.
 Range (min … max):  660.038 ns … 30.517 μs  ┊ GC (min … max):  0.00% … 96.61%
 Time  (median):     682.103 ns              ┊ GC (median):     0.00%
 Time  (mean ± σ):   825.723 ns ±  1.768 μs  ┊ GC (mean ± σ):  14.65% ±  6.64%

  ▅█▇▇▆▅▅▅▄▄▄▃▃▂▁▁▁▁                                           ▂
  ████████████████████▇████▇▇▇█▇▇▆▇▇▆▇▇▆▆▇██▇▇▇█▇▇▇▇▇▆▅▅▅▂▄▄▂▅ █
  660 ns        Histogram: log(frequency) by time      1.01 μs <

 Memory estimate: 1.31 KiB, allocs estimate: 15.
```

What do you think about?